Detect mips CPUs in ./configure

infinity0 opened the pull request:
This mirrors existing logic already in src/bootstrap/bootstrap.py

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

xen0n commented on configure:
```diff
@@ -388,6 +388,7 @@ msg "inspecting environment"
 
 CFG_OSTYPE=$(uname -s)
 CFG_CPUTYPE=$(uname -m)
+ENDIAN=$(printf '\1' | od -dAn)
```
Here `od` isn't `need_cmd`'d, thus unsafe to rely on. You could refer to the [`rustup-init.sh`] implementation for an alternative that only requires `head` and `tail`.

[`rustup-init.sh`]: https://github.com/rust-lang-nursery/rustup.rs/blob/1bedbb1b81a9a4cf6c8927db27a957e8b1fa18bb/rustup-init.sh#L125-L142

infinity0 replied:
`od` is [part of](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/od.html) POSIX (including the `-d` and `-A n` options), just as `head` and `tail` are. Is that good enough? If not, I can do your suggestion.

xen0n replied:
I don't know, but some exotic environments indeed doesn't have `od`; a quick Googling reveals OpenWRT and Maemo and the such. I don't know how many people are compiling Rust *on* these platforms though...

alexcrichton replied:
In order to minimize the possible impact of this change, could this move into just the mips block below?

infinity0 commented:
Welcome! In the meantime I spotted some more things which are missing:

- mk/cfg/mips{,el}-unknown-linux-gnu.mk: doesn't define CFG_RUN_TARG_mips{,el}-unknown-linux-gnu
- mk/cfg/mips64{,el}-unknown-linux-gnuabi64.mk are empty

I'll fix these as well and add them to this PR.

In the long run, does the rust team have any plans to unify the Makefiles and rustup/rustbuild? Seems like unnecessary work to have to maintain both of these...

xen0n commented:
@infinity0 Please refrain from touching the makefiles, they're [slated to be deleted pretty soon][deprecated]!

[deprecated]: https://internals.rust-lang.org/t/proposal-for-promoting-rustbuild-to-official-status/4368

infinity0 commented:
Ouch, OK. Is there a way to not use Cargo to build rust? That is going to make bootstrapping new architectures much much more difficult.

xen0n commented:
@infinity0 No, bootstraping will be arguably *easier*, since you no longer have to deal with multiple different build tools with rustbuild. Look at [the way `rustc` is `configure`'d for the non-x86 hosts][cross-rustbuild], pretty much *all* of the work you're expected to do for bootstrapping is now concentrated in the codebase itself. See #38314 for a feel of it; bootstrapping is lighter than ever IMO.

[cross-rustbuild]: https://buildbot.rust-lang.org/builders/nightly-dist-rustc-cross-rustbuild-linux/builds/84/steps/configure/logs/stdio

infinity0 commented:
Er, I am doing bootstrapping *literally right now*, and I can tell you it will be harder. All of the previous tools like `make` etc are already present for other architectures; cargo is not. Large cyclic dependencies also increase the amount of binary non-source code that needs to be trusted when bootstrapping a given whole distribution. From this perspective, it is better for rust to depend on 100 other build tools (that eventually depend on a small set of cyclic-dependency base build tools) than to be cyclicly dependent with cargo. I can appreciate the code is nicer to review than the previous Makefiles, but it would be nice if this could be done in a way that does not hard-depend on cargo. I made a concrete suggestion in #38652 and we can continue the discussion there.

sylvestre commented:
:+1: to what @infinity0 just said.

alexcrichton commented:
This PR looks good to me, thanks @infinity0!

infinity0 commented:
OK, I've moved `ENDIAN` down into the mips-specific block. I also added another commit that was necessary to compile the tests. Due to other problems (#39013, #39014) I couldn't test these all the way through to completion, however I notice that these files:
~~~~
src/tools/compiletest/src/util.rs
src/test/run-make/atomic-lock-free/Makefile
~~~~
mention `mips` but not `mips64`. Is that intentional or do they also need to be fixed?

alexcrichton commented:
@bors: r+ 

Thanks! It's ok to not update those unless necessary.

bors commented:
:pushpin: Commit 246e749 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 246e7492af6298e9ddf448c3c2aa67d41154f73a -->

bors commented:
:hourglass: Testing commit 246e749 with merge dd8e68d...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1522), [status-travis](https://travis-ci.org/rust-lang/rust/builds/191473845)
Approved by: alexcrichton
Pushing dd8e68d0da8f78b0aee367ddf154b3942940f3d3 to master...
